### PR TITLE
Introduce function-local settings for executor, expose in c++

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74348
* #74076
* #74186
* __->__ #74012
* #73938
* #73876
* #73875

This allows setting an executor on a function. The first use case is use to decompositions in C++ without additional fusion passes etc which might not work with custom tensors like batched tensors/vmap. A subsequent use case might be taking advantage of invokees of JIT execution which guard on certain properties before invocation (such as complete shapes in AOT autograd, rank in lazy tensor).

Differential Revision: [D34938124](https://our.internmc.facebook.com/intern/diff/D34938124)

### CODE REVIEW

File: torch/csrc/jit/runtime/simple_graph_executor_impl.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    c10::optional<size_t> remaining_bailout_depth) {
+  std::lock_guard<std::mutex> lock(compile_mutex);
+
+  // IMPORTANT: This is a hot path of calling a torchscript function. Try not to
+  // add any code above this.
+  if (execution_plan_) {
+    return *execution_plan_;
+  }
+  auto copy = graph->copy();
+  runNooptPassPipeline(copy);
```
Comment (Krovatkin) [inline torch/csrc/jit/runtime/simple_graph_executor_impl.cpp:27]: lmao why do we call this `Noopt` ? :)

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: torch/csrc/jit/api/function_impl.h [hunk 2]
```diff
@@ -13,10 +13,14 @@ struct TORCH_API GraphFunction : public Function {
...
       std::shared_ptr<Graph> graph,
-      std::function<void(GraphFunction&)> function_creator)
+      std::function<void(GraphFunction&)> function_creator,
+      c10::optional<ExecutorExecutionMode> executor_execution_mode =
+          c10::nullopt)
       : name_(std::move(name)),
         graph_(std::move(graph)),
-        function_creator_(std::move(function_creator)) {}
+        function_creator_(std::move(function_creator)) {
+    executor_execution_mode_ = executor_execution_mode;
```
Comment (Krovatkin) [inline torch/csrc/jit/api/function_impl.h:22]: why is this initialized inside a c-tor vs doing it like this: `graph_(std::move(graph)),`

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

**Comment by eellison** [comment]
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34938124).

**Pull request closed without merging**